File: java/std_nat/src/fan/std/AtomicRef.java
//
// Copyright (c) 2010, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Feb 10 Brian Frank  Creation
//
package fan.std;

import java.util.concurrent.atomic.AtomicReference;

import fan.sys.*;
import fanx.main.Sys;
import fanx.main.Type;

public final class AtomicRef extends FanObj {
	private static Type type = null;

	public Type typeof() {
		if (type == null) {
			type = Sys.findType("std::AtomicRef");
		}
		return type;
	}
	
	public AtomicRef(Object val) {
		if (!FanObj.isImmutable(val)) throw NotImmutableErr.make();
		this.val = new AtomicReference(val);
	}
	
	public static AtomicRef make() {
		AtomicRef self = new AtomicRef(null);
		return self;
	}

	public static AtomicRef make(Object val) {
		AtomicRef self = new AtomicRef(val);
		return self;
	}

	public final Object val() {
		return val.get();
	}
	
	public final Object get() {
		return val.get();
	}

	public final void val(Object v) {
		if (!FanObj.isImmutable(v))
			throw NotImmutableErr.make();
		val.set(v);
	}
	
	public final void set(Object v) {
		if (!FanObj.isImmutable(v))
			throw NotImmutableErr.make();
		val.set(v);
	}

	public final Object getAndSet(Object v) {
		if (!FanObj.isImmutable(v))
			throw NotImmutableErr.make();
		return val.getAndSet(v);
	}

	public final boolean compareAndSet(Object expect, Object update) {
		if (!FanObj.isImmutable(update))
			throw NotImmutableErr.make();
		return val.compareAndSet(expect, update);
	}
	
	public String toStr() {
		return val.toString();
	}

	private final AtomicReference val;
}

File: src/std/js/AtomicRef.js

fan.std.AtomicRef = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.AtomicRef.prototype.$ctor = function() {}
fan.std.AtomicRef.prototype.$typeof = function() { return fan.std.AtomicRef.$type; }

fan.std.AtomicRef.make = function(v) {
	self = new fan.std.AtomicRef();
	if (v === undefined) v = false;
	self.m_val = v;
	return self;
}

fan.std.AtomicRef.prototype.get = function() {
	return self.m_val;
}

fan.std.AtomicRef.prototype.val = function() {
	return self.m_val;
}

fan.std.AtomicRef.prototype.val$ = function(v) {
	self.m_val = v;
}

fan.std.AtomicRef.prototype.set$ = function(v) {
	self.m_val = v;
}

fan.std.AtomicRef.prototype.getAndSet = function(v) {
	o = this.m_val;
	this.m_val = v;
	return o;
}

fan.std.AtomicRef.prototype.compareAndSet = function(expect, update) {
	if (this.m_val == expect) {
		this.m_val = update;
		return true;
	}
	return false;
}
